lib/exla/op.ex

```diff
@@ -0,0 +1,39 @@
+defmodule Exla.Op do
+  alias __MODULE__, as: Op
+  alias Exla.Shape
+  alias Exla.Builder
+
+  @enforce_keys [:builder, :ref]
+  defstruct [:builder, :ref]
+
+  # The XLA API is explicit about the rank of the constant being created e.g. ConstantR0, ConstantR1
+  # We can be just as explicit, or we can use pattern matching on the inputs, I lean pattern matching
+  # as I think it makes the API feel more flexible
+  def constant(builder = %Builder{}, value) when is_number(value) do
+    case Exla.NIF.constant_r0(builder.ref, value) do
+      {:ok, ref} -> {:ok, %Op{builder: builder.ref, ref: ref}}
+      {:error, msg} -> {:error, msg}
+    end
+  end
+
+  def parameter(%Builder{ref: builder}, i, %Shape{ref: shape}, name)
+      when is_integer(i) and i >= 0 and is_binary(name) do
+    case Exla.NIF.parameter(builder, i, shape, name) do
+      {:ok, ref} -> {:ok, %Op{builder: builder, ref: ref}}
+      {:error, msg} -> {:error, msg}
+    end
+  end
+
+  # TODO: builder is redundant here because it's contained within each op
```

Comment:
You are right. We can probably skip the builder as long as the builder in both are equal!